Sources/SPMBuildCore/PluginInvocation.swift

```diff
@@ -52,95 +52,107 @@ extension PluginTarget {
         scriptRunner: PluginScriptRunner,
         outputDirectory: AbsolutePath,
         toolNamesToPaths: [String: AbsolutePath],
+        fileSystem: FileSystem,
         observabilityScope: ObservabilityScope,
-        fileSystem: FileSystem
-    ) throws -> PluginInvocationResult {
+        on queue: DispatchQueue,
+        completion: @escaping (Result<PluginInvocationResult, Error>) -> Void
+    ) {
         // Create the plugin working directory if needed (but don't do anything with it if it already exists).
         do {
             try fileSystem.createDirectory(outputDirectory, recursive: true)
         }
         catch {
-            throw PluginEvaluationError.outputDirectoryCouldNotBeCreated(path: outputDirectory, underlyingError: error)
+            return completion(.failure(PluginEvaluationError.couldNotCreateOuputDirectory(path: outputDirectory, underlyingError: error)))
         }
 
         // Create the input context to send to the plugin.
         // TODO: Some of this could probably be cached.
         var serializer = PluginScriptRunnerInputSerializer(buildEnvironment: buildEnvironment)
-        let inputStruct = try serializer.makePluginScriptRunnerInput(
-            rootPackage: package,
-            pluginWorkDir: outputDirectory,
-            builtProductsDir: outputDirectory,  // FIXME — what is this parameter needed for?
-            toolNamesToPaths: toolNamesToPaths,
-            pluginAction: action)
+        let inputStruct: PluginScriptRunnerInput
+        do {
+            inputStruct = try serializer.makePluginScriptRunnerInput(
+                rootPackage: package,
+                pluginWorkDir: outputDirectory,
+                builtProductsDir: outputDirectory,  // FIXME — what is this parameter needed for?
+                toolNamesToPaths: toolNamesToPaths,
+                pluginAction: action)
+        }
+        catch {
+            return completion(.failure(PluginEvaluationError.couldNotSerializePluginInput(underlyingError: error)))
+        }
         
         // Call the plugin script runner to actually invoke the plugin.
-        // TODO: This should be asynchronous.
         var outputText = Data()
-        let outputStruct = try scriptRunner.runPluginScript(
+        scriptRunner.runPluginScript(
             sources: sources,
             input: inputStruct,
             toolsVersion: self.apiVersion,
             writableDirectories: [outputDirectory],
+            fileSystem: fileSystem,
             observabilityScope: observabilityScope,
-            textOutputHandler: { data in
+            on: DispatchQueue(label: "plugin-invocation"),
```

Comment:
should this use '.sharedConcurrent'?

Reply:
I think that would result in text output from the plugin possibly arriving concurrently or out of order, and also not ensure that the completion block happens after all the output has arrived.

In the next phase I'm switching the block to a delegate that receives messages about plugin output and then eventual completion, and so the intent was to make sure those calls are ordered.  Very open to suggestions for alternate ways of doing that, but it sounds as if the callbacks should happen on a serial queue of some kind.

I patterned this after other places where there is an `on queue` parameter — that's just for the callbacks and completion blocks, right, as the actually processing is always on a shared queue?  Or is this queue parameter usually intended to be where the processing happens?

Reply:
the queue normally is for the callbacks, as the processing queue would normally be an internal implementation detail

Reply:
is this how the queue is used here too?

Reply:
Yes, exactly.

It might be more clear in the next PR where this is a delegate, if I change the name to `delegateQueue` as in manifest loader.  It's the same situation of wanting the "willDo" before the "didDo" call, type of thing.

Reply:
I started with it being called `callbackQueue` but changed it to match the others — I'll change it back.  I agree it's more clear.